File: Computer Architecture & Organization/decimal_to_binary/decimal_to_binary_cpp/NumberSystem.cpp
#include "NumberSystem.h"
#include <cmath>

int NumberSystem::decToBin(const int number) {
	int currentVal = number;
	int remainder = 0;
	int decimalPlace = 0;
	int output = 0;

	while (currentVal != 0) {
		remainder = currentVal % 2;
		currentVal /= 2;
		output += remainder * std::pow(10, decimalPlace);
		decimalPlace++;
	}

	return output;
}

double NumberSystem::decToBin(const double number) {
	int whole = number;
	double fractional = number - whole;
	double output = decToBin(whole);
	int decimalPlace = 1;

	// calculate fractional part
	while (fractional != 0) {
		whole = fractional * 2;
		fractional = (fractional * 2.0) - whole;
		output += whole * std::pow(10, -decimalPlace);
		decimalPlace++;
	}

	return output;
}

File: Computer Architecture & Organization/decimal_to_binary/decimal_to_binary_c#/NumberSystem.cs
namespace NumSysConvLib.Conversions
{
    /// <summary>
    /// A class full of functionalities for converting to different type of number systems.
    /// </summary>
    public static class NumberSystem
    {
        // Decimal to Binary
        public static int DecToBin(int number)
        {
            int multiplier = 1;
            int output = 0;

            while (number != 0)
            {
                int remainder = number % 2;
                number /= 2;
                output += remainder * multiplier;
                multiplier *= 10;
            }
            return output;
        }

        public static double DecToBin(double number)
        {
            double output = DecToBin((int)number);
            double fractional = number - (int)number;
            int decimalPlace = -1;

            while (fractional != 0.0)
            {
                int whole = (int)(fractional * 2);
                fractional = (fractional * 2) - whole;
                output += whole * Math.Pow(10, decimalPlace);
                decimalPlace--;
            }
            return output;
        }
    }
}
